Support upgrades for connect refactor

ndhanushkodi opened the pull request:
Before the connect refactor, service registration in Consul was managed by the lifecycle sidecar, which would re-register the service with Consul every 10s. Now, service registration is managed by Endpoints controller.

In order to support upgrades to the refactored Endpoints controller, we need Endpoints controller to NOT register or deregister any services managed by lifecycle sidecar. To do this, the annotation `consul.hashicorp.com/connect-inject-managed-by` is added to pods by the mutating webhook, so endpoints controller will ignore older services managed by lifecycle sidecar (legacy services) for service registration/deregistration. 

To make sure endpoints controller only deregisters services managed by endpoints controller, a meta key indicating the service is managed by endpoints controller is added to the Consul service registration. 

To support health checks for legacy services, the Endpoints controller will always update the healthcheck for any pod, whether it's managed by Endpoints controller or not.

How I've tested this PR:
* Helm install v0.31.1 with the following values
  * <details>
  </details>
* k apply -f static-client.yaml
  * <details>
  </details>
* k apply -f static-server.yaml
  * <details>
  </details>
* Helm install latest with the following values
  * <details>
  </details>
* Check that the services/checks still exist: 
```
k exec pod/nitya-consul-9hhn7 -- curl -s localhost:8500/v1/agent/services?ns=foo | jq .
k exec pod/nitya-consul-9hhn7 -- curl -s localhost:8500/v1/agent/checks?ns=foo | jq .
```
* Delete the pods of one of the static-client deployment
* Check that the services/checks still exist, and now the services will have metadata that they are managed by endpoints controller.
* Mark the static-server unhealthy and see that the check is updated even though it's a legacy service.

How I expect reviewers to test this PR:
Code review and if you can, the steps above.


Checklist:
- [x] Tests added
- [ ] CHANGELOG entry added (*HashiCorp engineers only, community PRs should not add a changelog entry*)

kschoche commented on connect-inject/endpoints_controller.go:
```diff
@@ -156,30 +162,42 @@ func (r *EndpointsController) Reconcile(ctx context.Context, req ctrl.Request) (
-						return ctrl.Result{}, err
-					}
+					if managedByEndpointsController {
+						//fmt.Println("here")
```
```suggestion
```

ndhanushkodi replied:
Thank you!!! Sorry I missed some cleanups!!

kschoche commented on connect-inject/endpoints_controller.go:
```diff
@@ -156,30 +162,42 @@ func (r *EndpointsController) Reconcile(ctx context.Context, req ctrl.Request) (
 						return ctrl.Result{}, err
 					}
+					checks, _ := client.Agent().Checks()
+					fmt.Printf("\n\nCHECKS: %+v\n", checks)
+					fmt.Printf("0:%+v\n", checks["service:pod1-service-created-sidecar-proxy:1"])
+					fmt.Printf("1:%+v\n", checks["service:pod1-service-created-sidecar-proxy:2"])
```
```suggestion
```

kschoche commented on connect-inject/endpoints_controller.go:
```diff
@@ -156,30 +162,42 @@ func (r *EndpointsController) Reconcile(ctx context.Context, req ctrl.Request) (
+						r.Log.Error(err, "failed to update health check status for service", "name", serviceName)
 						return ctrl.Result{}, err
 					}
+					checks, _ := client.Agent().Checks()
```
Assuming this one was also not meant to make it in?

kschoche commented on connect-inject/endpoints_controller.go:
```diff
@@ -210,6 +228,111 @@ func (r *EndpointsController) SetupWithManager(mgr ctrl.Manager) error {
+}
+
+// registerConsulHealthCheck registers a TTL health check for the service on this Agent.
+// The Agent is local to the Pod which has a kubernetes health check.
```
nit: the pod may not have a k8s health check, but it will have a Ready status which is what we're actually sync'ing.

ndhanushkodi replied:
Updated this to use the phrase "readiness status", WDYT?

kschoche replied:
Love it!

kschoche commented on connect-inject/endpoints_controller.go:
```diff
@@ -210,6 +228,111 @@ func (r *EndpointsController) SetupWithManager(mgr ctrl.Manager) error {
+		if err != nil {
+			return err
+		}
+	} else if serviceCheck.Status != status {
```
We could in theory always update, regardless of whether or not the status has changed, it is a TTL check after-all. It's a tradeoff between making an extra call and the code looking possibly cleaner?

Definitely not a blocker but I thought I'd raise the question in case anyone else thought it. I'm normally not a fan of issuing unnecessary IO ;)

ndhanushkodi replied:
I'd imagine that a large majority of the time the status doesn't change since people like to keep their apps up for a majority of the time and since it would avoid the extra call a majority of the time I'd lean towards leaving the check

kschoche replied:
Yup, however an update to the endpoints object is almost always driven by a Ready status change, but you're right that for most of the pods referenced by this endpoint object that we're reconciling it wouldnt need to be updated. i'm cool either way!

ndhanushkodi replied:
Oh yes that's true! I didn't think about that. Hmm, I think I still slightly lean towards keeping the check for a reduction in calls even though its not as many as I originally imagined.

kschoche replied:
Cool!

ndhanushkodi commented on connect-inject/endpoints_controller_test.go:
```diff
@@ -925,11 +927,321 @@ func TestReconcileUpdateEndpoint(t *testing.T) {
 	}{
+		// Legacy services are not managed by endpoints controller, but endpoints controller
+		// will still add/update the legacy service's health checks.
+		{
```
To reviewers since EP controller tests are long:
The tests I added were that for a legacy service: 
- the health check is added when the pod is healthy
- the health check is added when the pod is unhealthy
- the health check is updated from healthy --> unhealthy
- the health check is updated from unhealthy --> healthy

Since when an agent is rolled it will lose the health checks, I wanted to test that they are added back for a legacy service, and that they are updated if they already exist.

kschoche commented on connect-inject/endpoints_controller.go:
```diff
@@ -210,6 +223,111 @@ func (r *EndpointsController) SetupWithManager(mgr ctrl.Manager) error {
 
+// getServiceCheck will return the health check for this pod and service if it exists.
+func getServiceCheck(client *api.Client, healthCheckID string) (*api.AgentCheck, error) {
+	filter := fmt.Sprintf("CheckID == `%s`", healthCheckID)
```
I'm wondering if we may encounter a bug similar to the one where we weren't using `%q` when defining a filter in `connect-init` ? I'm guessing the `'s around "%s" make it work?

ndhanushkodi replied:
yeah I think so, and I didn't notice issues testing it end to end, and that codepath would be used for updating the health check which I did test :)

kschoche reviewed (approved):
This looks great!

lkysow reviewed (commented):
Sorry @ndhanushkodi. I couldn't complete this review. I removed myself from reviewers and added the plat team back and @ishustava was assigned.

I didn't get to finish but overall my concerns are that we're intertwining the code paths for endpoints-controller managed pods and legacy pods and that I'm worried that this is going to set us up for confusion.

I was looking to see if there was a clean way at the top of the `for` loop to deal with the legacy pod and then continue the loop without carrying the `isRegisteredByEndpointsCtrl` boolean throughout the rest of the loop code. I think that someone will add something at the bottom of the loop and be unaware that the code is dealing with legacy pods.

lkysow reviewed (approved):
Approving for now to not block.

ndhanushkodi commented:
> Correct me if my understanding is wrong: we're not creating the check here because we're going to create it after the service register so the check also gets created for non-endpoints-ctrl pods?

Yes

>If so, why don't we create it here and then after serviceRegister if the service isn't managed by endpoints controller we create the check?

Addressed below.

> overall my concerns are that we're intertwining the code paths for endpoints-controller managed pods and legacy pods and that I'm worried that this is going to set us up for confusion.
>I was looking to see if there was a clean way at the top of the for loop to deal with the legacy pod and then continue the loop without carrying the isRegisteredByEndpointsCtrl boolean throughout the rest of the loop code. I think that someone will add something at the bottom of the loop and be unaware that the code is dealing with legacy pods.

@lkysow I'm addressing these comments here:

The reason why I made the health check registration separate from the service registration was so we could have logic along the lines of:

```
// logic to create local client for agent
if managedByEndpointsController{
// create the service registration (without the healthcheck)
// register the service
}
// idempotently upsert the healthcheck (legacy and endpoints controller managed pods)
```

I could see how a future contributor might add logic meant for endpoints-controller managed pods to the end of that, and have that apply to legacy and endpoints controller managed pods.  

One option to reduce confusion is with a comment describing that the code after the if managedByEndpointsController block applies to legacy and endpoints controller pods.  I will go ahead and do this option in this PR so when this is merged with the existing logic it's less confusing.

Then, in a future PR, we could make a refactor to have the logic look like this:

```
// logic to create local client for agent
if legacy{
// upsert healthcheck for legacy pod
} else {
// create service registration with healthcheck
// register the service
}
```

@ishustava and @kschoche I'm looking for feedback on whether you think I should go forward with the future refactor option after merging this PR with some more documentation. I'm also looking for feedback on that refactor idea.

ndhanushkodi commented on connect-inject/annotations.go:
```diff
@@ -102,6 +108,9 @@ const (
 	injected = "injected"
+
+	// endpointsController is the value for keyManagedBy.
+	endpointsController = "consul-k8s-endpoints-controller"
```
This is a great suggestion, changed this

ishustava commented:
@ndhanushkodi 

> Then, in a future PR, we could make a refactor to have the logic look like this:

I like the refactor idea to make it more clear.

ndhanushkodi commented:
Thanks @ishustava @kschoche !
